src/ore/src/tracing/optimizer.rs

```diff
@@ -0,0 +1,313 @@
+            let plan = plan.clone();
+            trace.push(plan, current_time)
+        }
+    });
+}
+
+/// A [`layer::Layer`] implementation for [`PlanTrace`].
+///
+/// Populates the `data` wrapped by the [`PlanTrace`] instance with
+/// [`TraceEntry`] values, one for each span with attached plan in its
+/// extensions map.
+impl<S, T> layer::Layer<S> for PlanTrace<T>
+where
+    S: subscriber::Subscriber + for<'span> tracing_subscriber::registry::LookupSpan<'span>,
+    T: 'static,
+{
+    fn on_new_span(&self, attrs: &span::Attributes<'_>, id: &span::Id, ctx: layer::Context<'_, S>) {
+        // push to path stack
+        let mut path = self.path.lock().expect("path shouldn't be poisoned");
+        let name = attrs
+            .get_str("path.segment")
+            .unwrap_or_else(|| ctx.span(id).expect("span").name().to_string());
+        path.deref_mut().push(name);
+    }
+
+    fn on_enter(&self, _id: &span::Id, _ctx: layer::Context<'_, S>) {
+        // push to time stack
+        let mut times = self.times.lock().expect("times shouldn't be poisoned");
+        times.deref_mut().push(std::time::Instant::now());
+    }
+
+    fn on_exit(&self, _id: &span::Id, _ctx: layer::Context<'_, S>) {
+        // pop from path stack
+        let mut path = self.path.lock().expect("path shouldn't be poisoned");
+        path.deref_mut().pop();
+        // pop from time stack
+        let mut times = self.times.lock().expect("times shouldn't be poisoned");
+        times.deref_mut().pop();
+    }
+}
+
+impl<T> PlanTrace<T> {
+    /// Create a new trace for plans of type `T`.
+    pub fn new() -> Self {
+        Self {
+            path: Mutex::new(PathBuf::new()),
+            times: Mutex::new(vec![]),
+            entries: Mutex::new(vec![]),
+        }
+    }
+
+    /// Drain the trace data collected so far.
+    ///
+    /// Note that this method will mutate the internal state of the enclosing
+    /// [`PlanTrace`] even though its receiver is not `&mut self`. This quirk is
+    /// required because the tracing `Dispatch` does not have `downcast_mut` method.
+    pub fn drain_as_vec(&self) -> Vec<TraceEntry<T>>
+    where
+        T: Debug,
+    {
+        let mut entries = self.entries.lock().expect("entries shouldn't be poisoned");
+        entries.split_off(0)
+    }
+
+    /// Push a trace entry for the given `plan` to the current trace.
+    ///
+    /// This is a noop if the call is within a context without an enclosing span.
+    fn push(&self, plan: T, time: std::time::Instant) {
+        let times = self.times.lock().expect("times shouldn't be poisoned");
+        if let Some(span_start) = times.last() {
+            let mut entries = self.entries.lock().expect("entries shouldn't be poisoned");
+            entries.push(TraceEntry {
+                instant: time,
+                duration: time.duration_since(*span_start),
+                path: self.current_path(),
+                plan,
```

Comment:
this kindof assume that each span only has 1 child right? or am i mis-reading

This requirement is fine, but we should either document it, or panic if we can detect it being violated

Reply:
Each span can have only one plan attached to it as an invariant and this plan should represent the end state of that span. This is documented as constraint (2) in the `trace_plan` function.

Spans can be nested in each other in a tree-like manner as with regular `tracing`.

Reply:
ah thats nice